node/src/components/consensus/protocols/highway/performance_meter.rs

```diff
@@ -0,0 +1,148 @@
+const BLOCKS_TO_CONSIDER: usize = 5;
+/// The average max quorum that triggers us to slow down: with this big or smaller average max
+/// quorum per `BLOCKS_TO_CONSIDER`, we increase our round length.
+const SLOW_DOWN_THRESHOLD: f64 = 0.8;
+/// The average max quorum that triggers us to speed up: with this big or larger average max quorum
+/// per `BLOCKS_TO_CONSIDER`, we decrease our round length.
+const ACCELERATION_THRESHOLD: f64 = 0.9;
+/// We will try to accelerate (decrease our round length) every `ACCELERATION_PARAMETER` rounds if
+/// we have a big enough average max quorum.
+const ACCELERATION_PARAMETER: u64 = 40;
+
+#[derive(DataSize, Debug, Clone)]
+pub(crate) struct PerformanceMeter {
+    our_vid: ValidatorIndex,
+    min_round_len: TimeDiff,
+    max_round_len: TimeDiff,
+    current_round_len: TimeDiff,
+    last_switch_round_id: Timestamp,
+}
+
+impl PerformanceMeter {
+    pub fn new(
+        our_vid: ValidatorIndex,
+        round_len: TimeDiff,
+        min_round_len: TimeDiff,
+        max_round_len: TimeDiff,
+        timestamp: Timestamp,
+    ) -> Self {
+        let current_round_id = state::round_id(timestamp, round_len);
+        Self {
+            our_vid,
+            min_round_len,
+            max_round_len,
+            current_round_len: round_len,
+            last_switch_round_id: current_round_id,
+        }
+    }
+
+    pub fn current_round_len(&self) -> TimeDiff {
+        self.current_round_len
+    }
+
+    /// Whenever a block is proposed in a round and nodes cast their confirmation and witness
+    /// units, a "max quorum" can be calculated for each node - the maximum quorum for which there
+    /// exists a level-1 summit within the round, containing the particular node.
+    /// This function calculates the average max quorum for this node out of `BLOCKS_TO_CONSIDER`
+    /// most recent ancestor proposals of the current fork choice.
+    /// If this average max quorum is below a certain threshold, a higher round length will be
+    /// returned.
+    /// If it is above a certain threshold, and the current round ID is divisible by a certain
+    /// number, a lower round length is returned.
+    pub fn calculate_new_length<C: Context>(&mut self, state: &State<C>) -> TimeDiff {
+        let panorama = state.panorama();
+        let latest_block = match state.fork_choice(panorama) {
+            Some(block) => block,
+            None => {
+                // we have no blocks to check - just return the current setting
+                return self.current_round_len;
+            }
+        };
+
+        let blocks_to_check = state.ancestor_hashes(latest_block);
+
+        let max_quora: Vec<_> = blocks_to_check
+            .take_while(|block| state.unit(block).round_id() >= self.last_switch_round_id)
+            .filter_map(|block| {
+                let round_id = state.unit(block).round_id();
+                (!matches!(
```

Comment:
Not really about the functionality, but what is the advantage of using a macro here instead of a regular match expression?

Reply:
It's just more compact.